### ISSUE TITLE
[Proposal] Add quantized tensor types.

### ISSUE THREAD

**Issue by mciprian13** [onnx/onnx#2659]
I know that there are some proposals of adding quantized types to the ONNX standard but I just want to clarify the approach. Myself I am contributing in a ML framework which supports ONNX, namely Glow which is an open source project: https://github.com/pytorch/glow. The pain right now is that ONNX is not mature enough in terms of quantization support to allow the description of pre-quantized models.
Notes:
- A quantized type is nothing but an integer type (int8, int16, int32) with the addition of 2 parameters which describe the translation to/from a float32 tensor:
  - **scale** (float32)
  - **offset** (normally chosen as int32 to accomodate all int8/int16/int32 quantized types) - this sometimes is also reffered to as "zero_point"
  - the correspondence between a float value **xf** and a quantized value **xq** is:
    **`xf = scale * (xq - offset)`**
- These quantized types can be named "int8q", "int16q", "int32q". Of course we can also talk about the uint8, uint16 or uint32 quantized types but these are not friendly to the hardware instructions since most instructions are using signed arithmetic. Might be one of the reason TensorFlowLite changed the infrastructure from uint8 to int8 quantization. Of course we can go lower than that, for example 4bit, 2bit or 1bit quantization (binary networks).
- The quantization schema does not require to be explicitly described in the tensors/operators. From my experience we can talk about the following quantization schemas:
  - **Asymmetric** - this means no assumptions/restrictions for the quantization parameters
  - **Symmetric** - this means that the quantization parameter **offset** is **0**. This has run-time benefits since the complexity of the operators is diminished (no **offset** subtraction is required at run-time)
  - **AsymmetricWithPower2Scale** - this means the **scale** quantization parameter is a power of 2 (e.g. 0.5 or 0.125 or 1.0 or 2.0) which also has some run-time benefits since some operations are resolved to up/down bitwise shifts.
  - **SymmetricWithPower2Scale** - this means the **offset** parameter is **0** and the **scale** parameter is a power of 2.
  - To be noted that different quantization schemas have different tradeoffs between arithmetic accuracy and run-time complexity.
- The thing I don't like is that it seems that some new ONNX operators define quantization support by embedding the quantization information in the operators themselves instead of the tensor types. Such examples are `ConvInteger` and `MatMulInteger` where some quantization information is embedded in attributes like **x_zero_point** or **w_zero_point**. I don't think this is the right approach since this means new operators will be defined just for the purpose of allowing quantized data (although the meaning of the operators is the same) while the quantized nature should be embedded in the tensor types.
- In my opinion the operator definition should remain the same while adding new tensor types for quantized data. The rest of the implementation details (how the operators translate to run-time implementations) should be left to the backends supporting the ONNX format. An example of quantized tensor type would be:
```
    type {
      tensor_type {
        elem_type: xx            // Identifier for the quantized types "int8q" or "int16q" or "int32q"
        scale: 0.23                // scale parameter
        offset: -10                // offset parameter
        shape {
          dim {
            dim_value: 5
          }
        }
      }
    }
```
- Worth mentioning is that a new degree of generalization should allow the so called `sub-tensor quantization`. The examples above assume all the values of a tensor are quantized using the same parameters (scale and offset). Some applications require the quantization of tensors with smaller granularity meaning that sub-parts of the tensor are quantized using different parameters for better encoding of the values. For example this mechanism is beneficial for the so-called depth-wise convolution where each slice of the weight tensor has an independent histogram is better suited for representation by using their own scale & offset. For example a weight matrix could be quantized row-wise meaning each row will be quantized separately, resulting in a matrix of int8 values but with different quantization parameters for each row. This more general case could also be included in the proto format by having a quantization axis (or axes) and a vector of scales and offsets.

I am very interested in the future of the ONNX format and I would like to talk more about this. 
What do you think?
It would be nice if I could have a technical discussion about this feature with someone.
Best regards.

**Comment by linkerzhang** [comment]
Thanks for the interest on ONNX and ONNX quantization support.

This was discussed for very long time and The ONNX quantization design was agreed and made in current way - having all quantization parameters (schemas) as operators' inputs, which should be either inputs or attributes.

Adding tensor types for quantization is not a good idea in current ONNX level. It will introduce more and more such types as there're more and more quantization schemas.

However, this topic may be revisited in a higher level than ONNX, and it can be discussed in onnx-mlir project, which means, a higher level (than ONNX) dialect may be able to designed to handle different kinds of quantization schema or even customized schema.

**Comment by mciprian13** [comment]
I`m not sure I understand. The quantization schema relates to the particular values of the scale & offset parameters and does not influence (at all) the tensor/operator definition.
I don`t see a good thing of defining different operators just because they have different values for the quantization parameters.

**Comment by linkerzhang** [comment]
quantization schema means the math - say, scale/offset means linear quantization, there must be other math ways.

What do you mean of defining different operators just because they have different values for the quantization parameters? I don't think the quantization ops defined can't handle different scale/offset.

**Comment by linkerzhang** [comment]
For quantization ops defined, it's very clear that,
1. QLinear* ops were defined for linear quantization (with scale/offset math covered in the op definition).
2. *Integer ops were defined as more primitive ones handling 8-bit integers, without any quantization math included in the op definition. It means it can be used to compose any other math later.

As said, in ONNX level, it's not a good idea to define such data types, however, it may be good to define sth higher level than ONNX.

**Comment by mciprian13** [comment]
The scale/offset parameters are parameters like any other (is just like the tensor shape/size and type). That does not mean you need to define other operators types for different values. For example you don't define a `MatrixMultiplication2x2` operator type for 2x2 matrices and another operator type `MatrixMultiplication4x2` for 4x4 matrices. The operator type is the same but it can handle different matrix descriptions.
1. QLinear* - can handle linear quantization but it has the following drawbacks:
  - it assumes that all the operands are quantized. Why not allow mixed precision (float/fixed point) by making the quantization parameters part of the tensor description?
  - if ONNX has currently (e.g. say) 100 operators defined for float32 operations, does that mean that another definitions will be created for each of the 100 operators for quantized operation (say QLinearRNN, QLinearSqrt, ...). That is insane!
2. *Integer - These should be viewed as particular examples of the more generic quantized versions. For example `MatMulInteger` should be viewed as a particular instance of a more generic quantized Matrix Multiplication with the scale paramters (for A and B) equal to 1.0.

For other quantization "maths" (for example non-linear quantization) I don`t see why the quantization math cannot be described in the tensor itself.
Quantization is related to the data compression and not the operator definition.

Why not follow the path of TensorFlowLite? The quantization details are part of the tensor description and not part of the operators!

**Comment by linkerzhang** [comment]
All the points have been discussed very long time ago. most of the notes/agreements were summarized here https://github.com/onnx/onnx/wiki/Quantization-Support-In-ONNX.

For the points you made,
1. QLinear* ops are asking for all inputs quantized, that is true, because this was the most common quantization schema used and ONNX would standardize them as ops. For all others, ONNX is open enough, either they can be supported by customized ops/functions (with customers' own domain/version) or they can be proposed to ONNX community and be discussed.
2. *Integer ops are not taking any assumption on whether the inputs are quantized or not. They're doing regular maths against integers. For example, the inputs may not be quantized, but If the inputs are quantized, it's the caller's responsibility to ensure the inputs are sharing the same quantization parameters.

Current ONNX quantization solution supports both static and dynamic quantization very well. For static quantization, the quantization parameters can be taken as "constants" in the model. For dynamic quantization, they're very clearly taken as inputs. If there's any case not supported, the cases should be discussed too.

Furthermore, the discussion point here is actually about the direction of ONNX. Should ONNX support more and more advanced data structure, besides regular tensors? Should ONNX support more and more advanced operators?

I think it may be "YES" for both questions above. However,
1.  Supporting both above should be taken as layers. That means, not all advanced data structures or operators should be defined as one-layer common set in ONNX.
2. For operators, ONNX must have a very strict common primitive set of operators, but allow to have "functions" (advanced operators) so that it's simple to support more models (otherwise, a model will look to be very complex).
2. For data structures, ONNX does not have a good mechanism designed to allow extension to support advanced ones. This should be discussed - current MLIR framework may offer a way of thinking on it - to have operators/data types taken as multiple dialects (but layered on each other).

**Comment by sousoux** [comment]
We are now supporting ONNX as an input format for our toolchain https://github.com/GreenWaves-Technologies/gap_sdk/tree/master/tools/nntool. Having spent a lot of time getting quantization support working well with TFLITE, whether imported or determined by our tool, I would say that the comments above on tensor scale and offset annotation are correct. However I would go one step further. If ONNX is an export/import format (and not primarily a runner format) then the most important element of quantization is missing. That is not the ability to directly run the ONNX model quantized but to be able to determine the statistics on the activations without any training data. If ONNX was just to add min/mix/av/std information in all variable tensors that would go 100% of the way for allowing anyone to implement the quantization scheme that they wanted to. If ONNX supported that and some type of function annotation that at least preserves meta information about what a collection of operators was decomposed from then ONNX would (at least for us) become the export format of choice. I'm happy to participate in a working group to flesh this out if anyone is interested. Neither of these points involve changing how any of the ONNX operators work - just the addition of meta info that helps people producing backends to better map onto our optimized operators.
